File: archive/code/src/harness/providers/codex/usage.ts
import type { AgentUsage } from "../../events.js";
import {
  asRecord,
  numberField,
  pruneUndefined,
} from "./fields.js";

export function parseCodexAppServerUsage(value: unknown): AgentUsage | undefined {
  const usage = asRecord(value);
  const last = asRecord(usage.last);
  const total = asRecord(usage.total);
  const direct = parseCodexUsage(last);
  if (direct === undefined) return undefined;
  return pruneUndefined({
    ...direct,
    totalTokens:
      numberField(last, "totalTokens") ??
      numberField(last, "total_tokens") ??
      direct.totalTokens,
    totalProcessedTokens:
      numberField(total, "totalTokens") ?? numberField(total, "total_tokens"),
    maxTokens:
      numberField(usage, "modelContextWindow") ??
      numberField(usage, "model_context_window") ??
      null,
  });
}

function parseCodexUsage(value: unknown): AgentUsage | undefined {
  if (value === null || typeof value !== "object") return undefined;
  const obj = value as Record<string, unknown>;
  const inputTokens =
    numberField(obj, "input_tokens") ?? numberField(obj, "inputTokens");
  const cachedInputTokens =
    numberField(obj, "cached_input_tokens") ??
    numberField(obj, "cachedInputTokens") ??
    numberField(obj, "cacheReadTokens");
  const outputTokens =
    numberField(obj, "output_tokens") ?? numberField(obj, "outputTokens");
  const reasoningOutputTokens =
    numberField(obj, "reasoning_output_tokens") ??
    numberField(obj, "reasoningOutputTokens");
  return pruneUndefined({
    inputTokens,
    cachedInputTokens,
    outputTokens,
    reasoningOutputTokens,
    totalTokens:
      inputTokens !== undefined || outputTokens !== undefined
        ? (inputTokens ?? 0) + (outputTokens ?? 0)
        : undefined,
  });
}

File: src/codealmanac/integrations/harnesses/codex/usage.py
from pydantic import JsonValue

from codealmanac.engine.harnesses.models import HarnessUsage
from codealmanac.integrations.harnesses.codex.fields import (
    as_record,
    first_present,
    number_field,
)


def parse_codex_app_server_usage(value: JsonValue | None) -> HarnessUsage | None:
    usage = as_record(value)
    last = as_record(usage.get("last"))
    total = as_record(usage.get("total"))
    direct = parse_codex_usage(last)
    if direct is None:
        return None
    return direct.model_copy(
        update={
            "total_tokens": first_present(
                number_field(last, "totalTokens"),
                number_field(last, "total_tokens"),
                direct.total_tokens,
            ),
            "total_processed_tokens": first_present(
                number_field(total, "totalTokens"),
                number_field(total, "total_tokens"),
            ),
            "max_tokens": first_present(
                number_field(usage, "modelContextWindow"),
                number_field(usage, "model_context_window"),
            ),
        }
    )


def parse_codex_usage(value: JsonValue | None) -> HarnessUsage | None:
    obj = as_record(value)
    if len(obj) == 0:
        return None
    input_tokens = first_present(
        number_field(obj, "input_tokens"),
        number_field(obj, "inputTokens"),
    )
    cached_input_tokens = first_present(
        number_field(obj, "cached_input_tokens"),
        number_field(obj, "cachedInputTokens"),
        number_field(obj, "cacheReadTokens"),
    )
    output_tokens = first_present(
        number_field(obj, "output_tokens"),
        number_field(obj, "outputTokens"),
    )
    reasoning_output_tokens = first_present(
        number_field(obj, "reasoning_output_tokens"),
        number_field(obj, "reasoningOutputTokens"),
    )
    total_tokens = None
    if input_tokens is not None or output_tokens is not None:
        total_tokens = (input_tokens or 0) + (output_tokens or 0)
    return HarnessUsage(
        input_tokens=input_tokens,
        cached_input_tokens=cached_input_tokens,
        output_tokens=output_tokens,
        reasoning_output_tokens=reasoning_output_tokens,
        total_tokens=total_tokens,
    )
